### PR TITLE
Add missing define armadillo when use_extern_rng option is set to True

### PR DESCRIPTION
**Pull request by darcamo**
Specify library name and version:  **armadillo/10.7.0**

Add the define `ARMA_USE_EXTERN_RNG` when the `use_extern_rng` option is set to `True`, as in [my comment in issue #7334](https://github.com/conan-io/conan-center-index/pull/7334#issuecomment-951133407).

This is probably not necessary when the `use_wrapper` option is also set to `True`, since CMake should take care of that. However, since currently we have link problems when `use_wrapper` is set to `True` (see issue  [issue #7840](https://github.com/conan-io/conan-center-index/issues/7840)), I don't have a way to test it. Thus, I decided to be safe and always add the define when `use_extern_rng`  is set to `True`, regardless of the value of `use_wrapper`.

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by samuel-emrys** [comment]
You should be able to test this with the following command:

```bash
conan create . 10.7.0@user/experimental --build=missing =o armadillo:use_wrapper=True
```

You shouldn't get any linker errors with this command. Could you also add your snippet to `test_package/example.cpp` so that we can confirm that your change is making a difference. Something like:

```cpp
#ifdef ARMA_USE_EXTERN_RNG
    cout << "ARMA_USE_EXTERN_RNG set." << endl;
#else
    cout << "ARMA_USE_EXTERN_RNG not set." << endl;
#endif

    arma::vec v{1,2,3,4};
    v.print("v");

    arma::arma_rng::set_seed(1237);
    v.randn();
    v.print("v");
```

**Comment by darcamo** [comment]
I have added the change to the test code as an ammend to the previous commit.

I have then tested the recipe with 4 different configurations, varying `armadillo:use_wrapper` and `armadillo:use_extern_rng` values. 

With `armadillo:use_wrapper` set to `False` everything works as intended. I get the "ARMA_USE_EXTERN_RNG set." message when passing `armadillo:use_extern_rng=True` and "ARMA_USE_EXTERN_RNG not set." when passing `armadillo:use_extern_rng=True`.

However, when `armadillo:use_wrapper` is set to `True`, it always use external rnd and I get the "ARMA_USE_EXTERN_RNG set." message, regardless of the value of `armadillo:use_extern_rng=True`.

This should not be related to the code I have added to the recipe, though. It is probably armadillo's CMake code "being to smart" and enabling `ARMA_USE_EXTERN_RNG` when it detects some condition.

The values in the vector are also consistent, such that the same numbers are returned when we get the "ARMA_USE_EXTERN_RNG set."  message, and different ones when we get the "ARMA_USE_EXTERN_RNG not set." message.

File: recipes/armadillo/all/test_package/example.cpp
Comment (samuel-emrys) [inline recipes/armadillo/all/test_package/example.cpp:169]:
```suggestion
  arma::arma_rng::set_seed(1237);
  v2.randn();
  v2.print("v2 (randn):");
```

**Review by samuel-emrys — CHANGES_REQUESTED** [review]
Looking good. Small change for consistency with the way outputs are presented in `example.cpp`. Aside from that, there are a couple of things I think need to be added for validation. Looking at the library `CMakeLists.txt`, I found the following excerpts:

```cpp
# NOTE: ARMA_USE_EXTERN_RNG requires compiler support for thread_local and C++11
# NOTE: for Linux, this is available with gcc 4.8.3 onwards
# NOTE: for macOS, thread_local is supoported in Xcode 8 (mid 2016 onwards) in C++11 mode

# NOTE: thread_local appears broken again on macOS 11 (Big Sur) and/or AppleClang 12.0
# NOTE: see comments in include/armadillo_bits/arma_rng.hpp
...
if(MINGW OR MSYS OR CYGWIN OR MSVC)
  # MinGW doesn't correctly handle thread_local
  set(ARMA_USE_EXTERN_RNG false)
endif()
```
So perhaps the following changes are in order:
* Throw validation error when the condition `if use_wrapper and not use_extern_rng` is satisfied (add to `validate()`) since the external RNG _must_ be used when using the wrapper
* Add `del use_extern_rng` to `config_options()` when the user is using mingw, msys, cygwin or msvc

We could also make the compiler requirements explicit based on the logic presented in the library `CMakeLists.txt` - i.e., validation error on apple-clang12, gcc > 4.8.3, clang > 6.0, apple clang > 8.0 but I'm not sure if double handling these requirements is the best thing to do in the recipe. I'd say wait for feedback from one of the maintainers before making this change

**Review by SSE4 — APPROVED** [review]

**Comment by SSE4** [comment]
seems like it hard-codes absolute paths somewhere... package is not relocatable?

**Comment by samuel-emrys** [comment]
> seems like it hard-codes absolute paths somewhere... package is not relocatable?

Not that I can see?

**Comment by darcamo** [comment]
Indeed there is a new version, but I didn't want to make two unrelated changes in the same pull request. Thus, I have not changed the armadillo version in this pull request. If that one is merged before this one, I can rebase this pull request, if desired.

**Review by prince-chrismc — DISMISSED** [review]

**Comment by samuel-emrys** [comment]
> So perhaps the following changes are in order:
> * Throw validation error when the condition `if use_wrapper and not use_extern_rng` is satisfied (add to `validate()`) since the external RNG _must_ be used when using the wrapper
> * Add `del use_extern_rng` to `config_options()` when the user is using mingw, msys, cygwin or msvc

For clarity, I would still like to see these two changes added to this PR before it's merged in

**Comment by darcamo** [comment]
I added two commits, one that raises a `ConanInvalidConfiguration` exception if `use_wrapper` is True, but `use_extern_rng` is False, and another that deletes the `use_extern_rng` option when MINGW, MSYS, CYGWIN or MSVC are detected.

Unfortunately I don't have a way to test on windows, to check that the second commit is correct. From the documentation `conans.tools.os_info.detect_windows_subsystem` would tell me if any of MINGW, MSYS or CYGWIN is used, while MSVC could be checked with `self.settings.compiler` in the recipe. However, CI failed due to `tools.os_info` being forbidden in `config_options`. Can `self.settings.compiler` be used to detect the other values? (I could not find the possible values it can return in the documentation).

ps: I have the `CONAN_V2_MODE` environment variable defined, but I could build the recipe locally without any problems due to usage of `tools.os_info` in `config_options`.

File: recipes/armadillo/all/conanfile.py
Comment (samuel-emrys) [inline recipes/armadillo/all/conanfile.py:159]:
```suggestion
        # Ignore use_extern_rng when the option has been removed
        if self.options.use_wrapper and not self.options.get_safe("use_extern_rng", True):
            raise ConanInvalidConfiguration(
                "The wrapper requires the use of an external RNG. Set use_extern_rng=True and try again."
```

File: recipes/armadillo/all/conanfile.py
Comment (samuel-emrys) [inline recipes/armadillo/all/conanfile.py:282]:
```suggestion
        if self.options.get_safe("use_extern_rng"):
```

**Review by samuel-emrys — COMMENTED** [review]
Made a couple of changes to make this compatible with a windows build. I wasn't able to suggest a change for the necessary cmake defintions change though, so for this to work you'll need to make the following change:

```diff
-self._cmake.definitions[ARMA_USE_EXTERN_RNG"] = self.options.use_exern_rng
+self._cmake.definitions[ARMA_USE_EXTERN_RNG"] = self.options.get_safe("use_exern_rng", default=False)
```

**Comment by darcamo** [comment]
Importing `MSYS`, `MSYS2` and `CYGWIN` to use in the comparison is a bit better, since we don't hard code the compiler strings in the recipe. However, the problem of using `tools.os_info` in the `config_options` method, which CI is flagging as forbidden, remains.

**Comment by samuel-emrys** [comment]
If anybody could provide advice on how to proceed given the current issue, it would be appreciated. Is there an idiomatic way to query the windows subsystem in `config_options()`?

**Comment by prince-chrismc** [comment]
// cc @toge #8376 is similar?

**Comment by samuel-emrys** [comment]
It's not clear to me how the changes in #8376 are supposed to work - running the following:

```python
print("Windows Subsystem in Settings: {}".format(self.settings.os.subsystem))
print("Windows Subsystem in os_info: {}".format(tools.os_info.detect_windows_subsystem()))
```
yields
```
Windows Subsystem in Settings: None
Windows Subsystem in os_info: msys
```

Is the idea that the user sets `settings.os.subsystem` in their profile? This would feel like a bit of a clunky approach since it doesn't detect the environment

**Comment by prince-chrismc** [comment]
Sadly, I'm no expert on cross compiling environments

**Comment by toge** [comment]
@samuel-emrys 
Hello. I'am author of #8376.
I think your comments are right.  
Specifying `settings.os.subsystem` seems not to be right way.

I'm waiting for answer on your issues, too.
https://github.com/conan-io/conan-center-index/issues/8123

**Comment by samuel-emrys** [comment]
For completeness, the [reason this error exists](https://github.com/conan-io/conan-center-index/issues/8123#issuecomment-991938825) is because
> Because this method might be called on another machine than the build machine, and it shouldn't have any effect on its behavior.

If the requirement is that we remove `use_extern_rng` when the subsystem is `mingw`, `msys`, `cygwin` or `msvc`, are these just analogs for "Windows"? Are there any valid windows subsystem for this build?

If not, perhaps it would be worth using that as the condition, i.e.

```python
if self.settings.os == "Windows":
    del self.options.use_extern_rng
```
I'm not clear if this would preclude this option when building with WSL though.

The other alternative is to exclude this validation altogether and let the build recipe silently set this option to false. I don't love that idea - you would get two different package id's for substantively the same build. We could also make it a loud failure by patching 

```cmake
if(MINGW OR MSYS OR CYGWIN OR MSVC)
  # MinGW doesn't correctly handle thread_local
  set(ARMA_USE_EXTERN_RNG false)
endif()
```

to something like

```cmake
if(MINGW OR MSYS OR CYGWIN OR MSVC)
  # MinGW doesn't correctly handle thread_local
  message(FATAL_ERROR "ARMA_USE_EXTERN_RNG cannot be true when built with MINGW, MSYS, CYGWIN or MSVC")
endif()
```
Any thoughts?

**Comment by SpaceIm** [comment]
I don't remember, what are we talking about?

**Comment by samuel-emrys** [comment]
@SpaceIm I raised https://github.com/conan-io/conan-center-index/issues/8123 because we've run into an issue where hooks are throwing errors because `tools.os_info.detect_subsystem()` is being used in `config_options()`. You provided the advice above indicating that this hook is there

> Because this method might be called on another machine than the build machine, and it shouldn't have any effect on its behavior.

The reason it's being used is because one of the options in this recipe, `use_extern_rng`, is not valid when the subsystem being used is mingw, msys, cygwin or msvc because they don't handle `thread_local` appropriately, according to the upstream build system. Therefore, when these subsystems are detected this recipe attempts to remove `use_extern_rng` as an option that can be selected. I've proposed a possible workaround above, but I don't think it's perfect. If you have any advice to progress forward it would be appreciated.

**Comment by SpaceIm** [comment]
> mingw, msys, cygwin or msvc

So Windows?

File: recipes/armadillo/all/conanfile.py
Comment (SpaceIm) [inline recipes/armadillo/all/conanfile.py:100]: honestly, just remove this option in the condition `if self.settings.os == "Windows"` above.

Comment (samuel-emrys) [reply recipes/armadillo/all/conanfile.py:100]: This was the thought I had [above](https://github.com/conan-io/conan-center-index/pull/7919#issuecomment-999229047). My uncertainty was how WSL would be treated in a case like this. If that's not an issue, I'm inclined to agree to simplify this to a check for Windows.

Comment (prince-chrismc) [reply recipes/armadillo/all/conanfile.py:100]: Let's go simple, if someone runs into issues they'll probably know more and we can always revisit this

File: recipes/armadillo/all/conanfile.py
Comment (samuel-emrys) [inline recipes/armadillo/all/conanfile.py:99]:
```suggestion
        if self.settings.os == "Windows":
```

**Review by samuel-emrys — CHANGES_REQUESTED** [review]
As discussed previously, @darcamo you will also need to modify line 197 as follows:

```diff
-self._cmake.definitions["ARMA_USE_EXTERN_RNG"] = self.options.use_exern_rng
+self._cmake.definitions["ARMA_USE_EXTERN_RNG"] = self.options.get_safe("use_exern_rng", default=False)
```

**Review by prince-chrismc — APPROVED** [review]

**Comment by prince-chrismc** [comment]
dont forget the suggestion above https://github.com/conan-io/conan-center-index/pull/7919#pullrequestreview-852570560

**Comment by darcamo** [comment]
After everything is approved, should I squash all commits into one?

**Review by samuel-emrys — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]

**Comment by prince-chrismc** [comment]
> After everything is approved, should I squash all commits into one?

Nop! That will dismiss all the reviews and force us to restart the review (github will lose our progress)

The bot does an auto squash merge, you can see the docs here https://github.com/conan-io/conan-center-index/blob/master/docs/review_process.md#automatic-merges for more

**Review by uilianries — APPROVED** [review]
LGTM

**Review by SSE4 — APPROVED** [review]

**Comment by SSE4** [comment]
> After everything is approved, should I squash all commits into one?

nope, it's not needed. bot will squish and merge itself.

**Pull request merged**
